File: Leetcode_problems/lam1.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class ReverseDiagonalSum {
    
    public static List<Integer> revdiag(int[][] mat, int srow, int scol) {
        List<Integer> delements = new ArrayList<>();
        int r = srow, c = scol;
        while (r < mat.length && c >= 0) {
            delements.add(mat[r][c]);
            r++;
            c--;
        }
        return delements;
    }

    public static void maxdiagsum(int[][] mat) {
        int M = mat.length;
        int N = mat[0].length;

        // Variables to track the maximum sum and corresponding diagonal
        int maxSum = Integer.MIN_VALUE;
        List<Integer> maxDiag = new ArrayList<>();

        // Check diagonals starting from each column of the first row
        for (int scol = 0; scol < N; scol++) {
            List<Integer> diag = revdiag(mat, 0, scol);
            int dsum = diag.stream().mapToInt(Integer::intValue).sum();
            if (dsum > maxSum) {
                maxSum = dsum;
                maxDiag = diag;
            }
        }

        // Check diagonals starting from each row of the last column
        for (int srow = 1; srow < M; srow++) {
            List<Integer> diag = revdiag(mat, srow, N - 1);
            int dsum = diag.stream().mapToInt(Integer::intValue).sum();
            if (dsum > maxSum) {
                maxSum = dsum;
                maxDiag = diag;
            }
        }

        // Output the results
        System.out.println(maxSum);
        Collections.sort(maxDiag, Collections.reverseOrder());
        for (int num : maxDiag) {
            System.out.print(num + " ");
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Read mat dimensions
        int M = sc.nextInt();
        int N = sc.nextInt();
        
        // Read mat
        int[][] mat = new int[M][N];
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                mat[i][j] = sc.nextInt();
            }
        }

        maxdiagsum(mat);
    }
}

File: Leetcode_problems/Python/lam1.py
def get_reverse_diagonal(matrix, start_row, start_col):
    diag_elements = []
    r, c = start_row, start_col
    while r < len(matrix) and c >= 0:
        diag_elements.append(matrix[r][c])
        r += 1
        c -= 1
    return diag_elements

def find_max_reverse_diagonal_sum(matrix):
    M = len(matrix)
    N = len(matrix[0])

    # Variables to track the maximum sum and corresponding diagonal
    max_sum = float('-inf')
    max_diag = []

    # Check diagonals starting from each column of the first row
    for start_col in range(N):
        diag = get_reverse_diagonal(matrix, 0, start_col)
        diag_sum = sum(diag)
        if diag_sum > max_sum:
            max_sum = diag_sum
            max_diag = diag

    # Check diagonals starting from each row of the last column
    for start_row in range(1, M):
        diag = get_reverse_diagonal(matrix, start_row, N - 1)
        diag_sum = sum(diag)
        if diag_sum > max_sum:
            max_sum = diag_sum
            max_diag = diag

    # Output the results
    print(max_sum)
    max_diag.sort(reverse=True)
    print(' '.join(map(str, max_diag)))

if __name__ == "__main__":
    # Example Input
    """
    Example Input:
    3 3
    1 2 3
    4 5 6
    7 8 9
    """

    # Read matrix dimensions
    M, N = map(int, input().split())
    
    # Read matrix
    matrix = [list(map(int, input().split())) for _ in range(M)]

    find_max_reverse_diagonal_sum(matrix)
